File: ktox-lua-cc/src/main/kotlin/programs/Digsite.kt
package programs

import common.turtleGetFuelLevel
import lib.CHEST_INTAKE_SLOT
import lib.IntSpan
import lib.Movement
import lib.calibrateMovement
import lib.cargoFull
import lib.dumpCargo
import lib.gpsLocate
import lib.headingDx
import lib.headingDz
import lib.navigateTo
import lib.pastEnd
import lib.restockChest
import lib.SHAPE_CIRCLE
import lib.SHAPE_RIGHT_TRIANGLE
import lib.SHAPE_TRIANGLE
import lib.shapeRowBounds
import lib.shapeRowCount
import lib.shapeSpineColumn
import lib.stepFor

// Footprint-driven excavator. Named "digsite" (not "excavate") to avoid
// clashing with CC:Tweaked's built-in turtle/excavate.lua program.
//
// Usage: digsite <width> (<length>) (<height>)
//        digsite -t <side> (<height>)
//        digsite -rt <side> (<height>)
//        digsite -c <radius> (<height>)
//        digsite -h
//   - width: footprint size along the turtle's right-hand side at start.
//   - length (optional, defaults to width — square footprint): footprint
//     size along the turtle's forward direction at start.
//   - height (optional, signed — positive climbs, negative descends): if
//     given, "room" mode — hollows out that exact bounded x/y/z volume
//     (homeY to homeY+height), one full horizontal layer at a time. If
//     omitted, "clear cut" mode — the turtle's own starting Y is the
//     FLOOR (never dug below); it sweeps full layers upward from there
//     through CLEAR_CUT_HEIGHT blocks (stopping early once a whole layer
//     digs nothing — see clearCut() below). Full-layer sweeps (not
//     per-column "dig until first gap") so overhangs/floating terrain get
//     cleared too.
//   All of width/length/height/side/radius are relative to home (the
//   turtle's start position/heading), like ExcavatePro's
//   width/length/depth, rather than absolute world coordinates — so this
//   works identically with or without GPS (see calibrateMovement() in
//   lib/Movement.kt). Replaces an earlier x1:x2/z1:z2
//   absolute-world-coordinate interface that broke (silently pointed at
//   meaningless/wrong ground) the moment GPS was unavailable.
//   Positional convention for the plain rectangle form: arg COUNT decides
//   meaning (1 = width only, 2 = width+length, 3 = width+length+height),
//   matching ExcavatePro, since there's no way to tell a bare length from
//   a bare height apart otherwise. -t/-rt/-c instead take a single
//   side/radius value in that slot (no separate "length" — see
//   lib/Shape.kt for what each shape looks like):
//     -t   isoceles triangle, base = <side>, apex pointing away from home.
//     -rt  right isoceles triangle (both legs = <side>), right-angle
//          vertex at home.
//     -c   filled circle of the given <radius>, centered <radius> blocks
//          right and forward of home.
//
// Fueling/storage: assumes a fuel chest directly behind the turtle's start
// position, with overflow chests extending from there. ASSUMPTION (easy to
// flip if wrong — see CHEST_OVERFLOW_SIDE in lib/Chest.kt): overflow chests
// extend toward the turtle's original right-hand side. Chest logic itself
// (dumping, restocking, the fuel allowlist) is shared with ExcavatePro via
// lib/Chest.kt — this file only decides which slot(s) to keep off-limits.

const val FUEL_SAFETY_MARGIN = 20

// Blocks swept upward from the floor in clear-cut mode. Raise this if your
// terrain runs taller than this above the turtle's starting position.
const val CLEAR_CUT_HEIGHT = 32

fun printUsage() {
    println("Usage: digsite <width> (<length>) (<height>)")
    println("       digsite -t <side> (<height>)   isoceles triangle")
    println("       digsite -rt <side> (<height>)  right triangle (equal legs)")
    println("       digsite -c <radius> (<height>) circle")
    println("       digsite -h                     show this help")
    println("height omitted: clear-cut mode, sweeps up from home Y until a")
    println("layer digs nothing. height given: room mode, exact bounded box.")
}

fun main(args: Array<String>) {
    if (args.size < 1) {
        printUsage()
        return
    }

    val flag = args[1]
    if (flag == "-h" || flag == "--help") {
        printUsage()
        return
    }

    val isShape = flag == "-t" || flag == "-rt" || flag == "-c"
    if (isShape && args.size < 2) {
        printUsage()
        return
    }

    println("Calibrating position via GPS...")
    val movement = calibrateMovement()
    if (movement.gpsEnabled) {
        println("Home at x=${movement.homeX} y=${movement.homeY} z=${movement.homeZ}")
    } else {
        println("No GPS available - running on dead reckoning only (no drift checks).")
    }

    val fwdDx = headingDx(movement.homeHeading)
    val fwdDz = headingDz(movement.homeHeading)
    val rightHeading = (movement.homeHeading + 1) % 4
    val rgtDx = headingDx(rightHeading)
    val rgtDz = headingDz(rightHeading)

    if (isShape) {
        val shape = shapeForFlag(flag)
        val size = args[2].toInt()
        val hasHeight = args.size >= 3
        val height = if (hasHeight) args[3].toInt() else 0

        val layerFn = { mv: Movement, y: Int ->
            digsiteLayerShaped(mv, movement.homeX, movement.homeZ, rgtDx, rgtDz, fwdDx, fwdDz, shape, size, y)
        }

        if (hasHeight) {
            val ySpan = IntSpan(movement.homeY, movement.homeY + height)
            digsiteRoom(movement, ySpan, layerFn)
        } else {
            clearCut(movement, layerFn)
        }
    } else {
        val width = args[1].toInt()
        val length = if (args.size >= 2) args[2].toInt() else width
        val hasHeight = args.size >= 3
        val height = if (hasHeight) args[3].toInt() else 0

        val xOther = movement.homeX + rgtDx * (width - 1) + fwdDx * (length - 1)
        val zOther = movement.homeZ + rgtDz * (width - 1) + fwdDz * (length - 1)
        val xSpan = IntSpan(movement.homeX, xOther)
        val zSpan = IntSpan(movement.homeZ, zOther)

        val layerFn = { mv: Movement, y: Int -> digsiteLayer(mv, xSpan, zSpan, y) }

        if (hasHeight) {
            val ySpan = IntSpan(movement.homeY, movement.homeY + height)
            digsiteRoom(movement, ySpan, layerFn)
        } else {
            clearCut(movement, layerFn)
        }
    }

    println("Excavation complete. Returning home...")
    navigateTo(movement, movement.homeX, movement.homeY, movement.homeZ)
    movement.faceHeading(movement.homeHeading)
    println("Home.")
}

fun shapeForFlag(flag: String): String {
    if (flag == "-t") {
        return SHAPE_TRIANGLE
    }
    if (flag == "-rt") {
        return SHAPE_RIGHT_TRIANGLE
    }
    return SHAPE_CIRCLE
}

// -- Base servicing (fuel + item disposal) --

// Only the shared restock scratch slot needs protecting — unlike
// ExcavatePro, there's no stock (torches, step material) to reserve a
// slot for, since fuel is consumed immediately by restockChest() rather
// than stockpiled.
fun keepSlot(slot: Int): Boolean {
    return slot == CHEST_INTAKE_SLOT
}

fun needsService(m: Movement): Boolean {
    val fuel = turtleGetFuelLevel()
    val distance = m.distanceHome()
    if (fuel < distance + FUEL_SAFETY_MARGIN) {
        return true
    }
    return cargoFull { slot -> keepSlot(slot) }
}

fun ensureFuelAndSpace(m: Movement) {
    if (needsService(m)) {
        serviceAtBase(m)
    }
}

fun serviceAtBase(m: Movement) {
    println("Returning to base to refuel/dump inventory...")
    val returnX = m.x
    val returnY = m.y
    val returnZ = m.z

    dumpCargo(m) { slot -> keepSlot(slot) }
    // Nothing else to sort here (no torches/stairs like ExcavatePro) —
    // any non-fuel item just means the chest doesn't hold what this
    // program expects, and restockChest will stop on it rather than spin.
    restockChest(m, 16) { name -> false }

    navigateTo(m, returnX, returnY, returnZ)

    // Drift safety check after the round trip — only meaningful when GPS
    // was available at calibration; otherwise there's no ground truth to
    // check against.
    if (m.gpsEnabled) {
        val checked = gpsLocate()
        if (checked != null) {
            m.x = checked.x
            m.y = checked.y
            m.z = checked.z
        }
    }
    println("Resuming excavation.")
}

// -- Room mode: bounded y-range excavation, one full layer at a time --
//
// The footprint itself (rectangle vs. a lib/Shape.kt shape) is entirely
// the caller's concern, via layerFn — this only walks the y-range and
// invokes it once per layer.

fun digsiteRoom(m: Movement, ySpan: IntSpan, layerFn: (Movement, Int) -> Unit) {
    val yStep = stepFor(ySpan)
    var y = ySpan.start
    while (!pastEnd(y, ySpan.finish, yStep)) {
        layerFn(m, y)
        y += yStep
    }
}

fun digsiteLayer(m: Movement, xSpan: IntSpan, zSpan: IntSpan, y: Int) {
    val xStep = stepFor(xSpan)
    val zStep = stepFor(zSpan)
    var x = xSpan.start
    var forwardZ = true
    while (!pastEnd(x, xSpan.finish, xStep)) {
        if (forwardZ) {
            var z = zSpan.start
            while (!pastEnd(z, zSpan.finish, zStep)) {
                ensureFuelAndSpace(m)
                navigateTo(m, x, y, z)
                z += zStep
            }
        } else {
            var z = zSpan.finish
            while (!pastEnd(z, zSpan.start, -zStep)) {
                ensureFuelAndSpace(m)
                navigateTo(m, x, y, z)
                z -= zStep
            }
        }
        forwardZ = !forwardZ
        x += xStep
    }
}

// -- Clear-cut mode: footprint (rectangle or shaped), floor = home Y,
// full layers upward --
//
// Sweeps full layers upward from home Y (the floor), one layerFn call per
// layer like digsiteRoom would over a fixed y-range, but bails out early
// once a whole layer dug nothing — that means the turtle has climbed
// above the terrain (open air/sky on every column), so grinding through
// the remaining CLEAR_CUT_HEIGHT layers just to confirm it's still empty
// wastes fuel/time for no benefit. Sweeping every layer in full (rather
// than following each column up until the first gap) is what actually
// clears overhangs/floating terrain — a gap below a column doesn't stop
// that column's higher blocks from being dug, since every other column
// at that same height gets visited regardless — the per-layer dug-count
// check preserves that: a layer only counts as "empty" when NONE of its
// columns dug anything.
//
// The floor layer itself is never skipped by this check even if it happens
// to dig nothing (e.g. the turtle started standing on bedrock/already-clear
// ground) — the loop always completes at least one full layer before the
// empty check can trigger a stop.

fun clearCut(m: Movement, layerFn: (Movement, Int) -> Unit) {
    val floorY = m.homeY
    val yStep = stepFor(IntSpan(floorY, floorY + CLEAR_CUT_HEIGHT))
    var y = floorY
    while (!pastEnd(y, floorY + CLEAR_CUT_HEIGHT, yStep)) {
        val dugBefore = m.blocksDug
        layerFn(m, y)
        if (m.blocksDug == dugBefore) {
            println("Layer at y=${y} dug nothing - assuming clear of terrain, stopping.")
            return
        }
        y += yStep
    }
}

// -- Shaped footprints (triangle/right-triangle/circle) --
//
// Sweeps one layer of a lib/Shape.kt footprint at the given world y. Each
// row (stepping along the forward axis from home) has its own included
// column range (shapeRowBounds) — the turtle transits between rows via
// the shape's spine column (shapeSpineColumn), which is guaranteed part
// of every row, then sweeps outward from the spine to each edge of the
// row's range and back. Never crosses a cell outside the shape, unlike a
// naive full-bounding-box sweep would.

fun digsiteLayerShaped(
    m: Movement,
    homeX: Int,
    homeZ: Int,
    rgtDx: Int,
    rgtDz: Int,
    fwdDx: Int,
    fwdDz: Int,
    shape: String,
    size: Int,
    y: Int,
) {
    val spine = shapeSpineColumn(shape, size)
    val rows = shapeRowCount(shape, size)
    var row = 0
    while (row < rows) {
        val bounds = shapeRowBounds(shape, size, row)
        if (bounds != null) {
            val spineX = homeX + rgtDx * spine + fwdDx * row
            val spineZ = homeZ + rgtDz * spine + fwdDz * row
            ensureFuelAndSpace(m)
            navigateTo(m, spineX, y, spineZ)

            var col = spine
            while (col > bounds.start) {
                col -= 1
                ensureFuelAndSpace(m)
                navigateTo(m, homeX + rgtDx * col + fwdDx * row, y, homeZ + rgtDz * col + fwdDz * row)
            }
            ensureFuelAndSpace(m)
            navigateTo(m, spineX, y, spineZ)

            col = spine
            while (col < bounds.finish) {
                col += 1
                ensureFuelAndSpace(m)
                navigateTo(m, homeX + rgtDx * col + fwdDx * row, y, homeZ + rgtDz * col + fwdDz * row)
            }
            ensureFuelAndSpace(m)
            navigateTo(m, spineX, y, spineZ)
        }
        row += 1
    }
}

File: src/pkg/player/8durt/Digsite.lua
-- package: programs

require("ktox-lib")
ktox_sourcemap_traceback(debug and debug.getinfo and (debug.getinfo(1) or {}).short_src or "", "Digsite.kt", {["1-15"]=1,["16"]=82,["17"]=83,["18"]=84,["19"]=85,["20"]=86,["21"]=87,["22-26"]=88,["27"]=92,["28"]=93,["29-30"]=94,["31"]=97,["32"]=98,["33"]=99,["34-35"]=100,["36"]=103,["37"]=104,["38"]=105,["39-40"]=106,["41"]=109,["42"]=110,["43"]=111,["44-45"]=112,["46-47"]=114,["48"]=117,["49"]=118,["50"]=119,["51"]=120,["52"]=121,["53"]=123,["54"]=124,["55"]=125,["56"]=126,["57"]=127,["58-60"]=129,["61"]=133,["62"]=134,["63-64"]=135,["65-67"]=137,["68"]=140,["69"]=141,["70"]=142,["71"]=143,["72"]=145,["73"]=146,["74"]=147,["75"]=148,["76-78"]=150,["79"]=152,["80"]=153,["81-82"]=154,["83-85"]=156,["86"]=160,["87"]=161,["88"]=162,["89-94"]=163,["95"]=167,["96-97"]=168,["98"]=170,["99-100"]=171,["101-106"]=173,["107-112"]=183,["113"]=187,["114"]=188,["115"]=189,["116-117"]=190,["118-124"]=192,["125"]=196,["126-134"]=197,["135"]=215,["136"]=216,["137-142"]=217,["143"]=222,["144"]=223,["145"]=224,["146"]=225,["147-149"]=227,["150-152"]=231,["153"]=233,["154"]=238,["155"]=239,["156"]=240,["157"]=241,["158"]=242,["159-161"]=243,["162-168"]=246,["169"]=256,["170"]=257,["171"]=258,["172"]=259,["173-181"]=260,["182"]=265,["183"]=266,["184"]=267,["185"]=268,["186"]=269,["187"]=270,["188"]=271,["189"]=272,["190"]=273,["191"]=274,["192-194"]=275,["195"]=278,["196"]=279,["197"]=280,["198"]=281,["199-201"]=282,["202"]=285,["203-209"]=286,["210"]=312,["211"]=313,["212"]=314,["213"]=315,["214"]=316,["215"]=317,["216"]=318,["217"]=319,["218-219"]=320,["220-234"]=322,["235"]=357,["236"]=358,["237"]=359,["238"]=360,["239"]=361,["240"]=362,["241"]=363,["242"]=364,["243"]=365,["244"]=366,["245"]=368,["246"]=369,["247"]=370,["248"]=371,["249-250"]=372,["251"]=374,["252"]=375,["253"]=377,["254"]=378,["255"]=379,["256"]=380,["257-258"]=381,["259"]=383,["260-261"]=384,["262-265"]=386}, "programs")
ktox_require("lib/Chest")
ktox_require("lib/Span")
ktox_require("lib/Movement")
ktox_require("lib/Position")
ktox_require("lib/Shape")

FUEL_SAFETY_MARGIN = 20

CLEAR_CUT_HEIGHT = 32

function printUsage()
    println("Usage: digsite <width> (<length>) (<height>)")
    println("       digsite -t <side> (<height>)   isoceles triangle")
    println("       digsite -rt <side> (<height>)  right triangle (equal legs)")
    println("       digsite -c <radius> (<height>) circle")
    println("       digsite -h                     show this help")
    println("height omitted: clear-cut mode, sweeps up from home Y until a")
    println("layer digs nothing. height given: room mode, exact bounded box.")
end

---@param args table
function main(args)
    if #(args) < 1 then
        printUsage()
        return
    end
    local flag = args[1]
    if flag == "-h" or flag == "--help" then
        printUsage()
        return
    end
    local isShape = flag == "-t" or flag == "-rt" or flag == "-c"
    if isShape and #(args) < 2 then
        printUsage()
        return
    end
    println("Calibrating position via GPS...")
    local movement = calibrateMovement()
    if movement.gpsEnabled then
        println("Home at x=" .. tostring(movement.homeX) .. " y=" .. tostring(movement.homeY) .. " z=" .. tostring(movement.homeZ))
    else
        println("No GPS available - running on dead reckoning only (no drift checks).")
    end
    local fwdDx = headingDx(movement.homeHeading)
    local fwdDz = headingDz(movement.homeHeading)
    local rightHeading = (movement.homeHeading + 1) % 4
    local rgtDx = headingDx(rightHeading)
    local rgtDz = headingDz(rightHeading)
    if isShape then
        local shape = shapeForFlag(flag)
        local size = ktox_toInt(args[2])
        local hasHeight = #(args) >= 3
        local height = (hasHeight and ktox_toInt(args[3]) or 0)
        local layerFn = function(mv, y)
            return digsiteLayerShaped(mv, movement.homeX, movement.homeZ, rgtDx, rgtDz, fwdDx, fwdDz, shape, size, y)
        end
        if hasHeight then
            local ySpan = IntSpan:new(movement.homeY, movement.homeY + height)
            digsiteRoom(movement, ySpan, layerFn)
        else
            clearCut(movement, layerFn)
        end
    else
        local width = ktox_toInt(args[1])
        local length = (#(args) >= 2 and ktox_toInt(args[2]) or width)
        local hasHeight = #(args) >= 3
        local height = (hasHeight and ktox_toInt(args[3]) or 0)
        local xOther = movement.homeX + rgtDx * (width - 1) + fwdDx * (length - 1)
        local zOther = movement.homeZ + rgtDz * (width - 1) + fwdDz * (length - 1)
        local xSpan = IntSpan:new(movement.homeX, xOther)
        local zSpan = IntSpan:new(movement.homeZ, zOther)
        local layerFn = function(mv, y)
            return digsiteLayer(mv, xSpan, zSpan, y)
        end
        if hasHeight then
            local ySpan = IntSpan:new(movement.homeY, movement.homeY + height)
            digsiteRoom(movement, ySpan, layerFn)
        else
            clearCut(movement, layerFn)
        end
    end
    println("Excavation complete. Returning home...")
    navigateTo(movement, movement.homeX, movement.homeY, movement.homeZ)
    movement:faceHeading(movement.homeHeading)
    println("Home.")
end

---@param flag string
---@return string
function shapeForFlag(flag)
    if flag == "-t" then
        return SHAPE_TRIANGLE
    end
    if flag == "-rt" then
        return SHAPE_RIGHT_TRIANGLE
    end
    return SHAPE_CIRCLE
end

---@param slot number
---@return boolean
function keepSlot(slot)
    return slot == CHEST_INTAKE_SLOT
end

---@param m Movement
---@return boolean
function needsService(m)
    local fuel = turtle.getFuelLevel()
    local distance = m:distanceHome()
    if fuel < distance + FUEL_SAFETY_MARGIN then
        return true
    end
    return cargoFull(function(slot)
        return keepSlot(slot)
    end)
end

---@param m Movement
function ensureFuelAndSpace(m)
    if needsService(m) then
        serviceAtBase(m)
    end
end

---@param m Movement
---@param spineX number
---@param y number
---@param spineZ number
function ensureFuelAndSpaceShaped(m, spineX, y, spineZ)
    if needsService(m) then
        navigateTo(m, spineX, y, spineZ)
        serviceAtBase(m)
    end
end

---@param m Movement
function serviceAtBase(m)
    println("Returning to base to refuel/dump inventory...")
    local returnX = m.x
    local returnY = m.y
    local returnZ = m.z
    dumpCargo(m, function(slot)
        return keepSlot(slot)
    end)
    restockChest(m, 16, function(name)
        return false
    end)
    navigateTo(m, returnX, returnY, returnZ)
    if m.gpsEnabled then
        local checked = gpsLocate()
        if checked ~= nil then
            m.x = checked.x
            m.y = checked.y
            m.z = checked.z
        end
    end
    println("Resuming excavation.")
end

---@param m Movement
---@param ySpan IntSpan
---@param layerFn (Movement, Int) -> Unit
function digsiteRoom(m, ySpan, layerFn)
    local yStep = stepFor(ySpan)
    local y = ySpan.start
    while not pastEnd(y, ySpan.finish, yStep) do
        layerFn(m, y)
        y = ktox_plusAssign(y, yStep)
    end
end

---@param m Movement
---@param xSpan IntSpan
---@param zSpan IntSpan
---@param y number
function digsiteLayer(m, xSpan, zSpan, y)
    local xStep = stepFor(xSpan)
    local zStep = stepFor(zSpan)
    local x = xSpan.start
    local forwardZ = true
    while not pastEnd(x, xSpan.finish, xStep) do
        if forwardZ then
            local z = zSpan.start
            while not pastEnd(z, zSpan.finish, zStep) do
                ensureFuelAndSpace(m)
                navigateTo(m, x, y, z)
                z = ktox_plusAssign(z, zStep)
            end
        else
            local z = zSpan.finish
            while not pastEnd(z, zSpan.start, -zStep) do
                ensureFuelAndSpace(m)
                navigateTo(m, x, y, z)
                z = ktox_minusAssign(z, zStep)
            end
        end
        forwardZ = not forwardZ
        x = ktox_plusAssign(x, xStep)
    end
end

---@param m Movement
---@param layerFn (Movement, Int) -> Unit
function clearCut(m, layerFn)
    local floorY = m.homeY
    local yStep = stepFor(IntSpan:new(floorY, floorY + CLEAR_CUT_HEIGHT))
    local y = floorY
    while not pastEnd(y, floorY + CLEAR_CUT_HEIGHT, yStep) do
        local dugBefore = m.blocksDug
        layerFn(m, y)
        if m.blocksDug == dugBefore then
            println("Layer at y=" .. tostring(y) .. " dug nothing - assuming clear of terrain, stopping.")
            return
        end
        y = ktox_plusAssign(y, yStep)
    end
end

---@param m Movement
---@param homeX number
---@param homeZ number
---@param rgtDx number
---@param rgtDz number
---@param fwdDx number
---@param fwdDz number
---@param shape string
---@param size number
---@param y number
function digsiteLayerShaped(m, homeX, homeZ, rgtDx, rgtDz, fwdDx, fwdDz, shape, size, y)
    local spine = shapeSpineColumn(shape, size)
    local rows = shapeRowCount(shape, size)
    local row = 0
    while row < rows do
        local bounds = shapeRowBounds(shape, size, row)
        if bounds ~= nil then
            local spineX = homeX + fwdDx * row
            local spineZ = homeZ + fwdDz * row
            ensureFuelAndSpaceShaped(m, spineX, y, spineZ)
            navigateTo(m, spineX, y, spineZ)
            local col = spine
            while col > bounds.start do
                col = ktox_minusAssign(col, 1)
                ensureFuelAndSpaceShaped(m, spineX, y, spineZ)
                navigateTo(m, homeX + rgtDx * (col - spine) + fwdDx * row, y, homeZ + rgtDz * (col - spine) + fwdDz * row)
            end
            ensureFuelAndSpaceShaped(m, spineX, y, spineZ)
            navigateTo(m, spineX, y, spineZ)
            col = spine
            while col < bounds.finish do
                col = ktox_plusAssign(col, 1)
                ensureFuelAndSpaceShaped(m, spineX, y, spineZ)
                navigateTo(m, homeX + rgtDx * (col - spine) + fwdDx * row, y, homeZ + rgtDz * (col - spine) + fwdDz * row)
            end
            ensureFuelAndSpaceShaped(m, spineX, y, spineZ)
            navigateTo(m, spineX, y, spineZ)
        end
        row = ktox_plusAssign(row, 1)
    end
end


main({...})
